### PR TITLE
Created Unit Tests and Fixed Errors

### PR DESCRIPTION
**Pull request by violet-moon**
Fixed the error where double nested config statements were not able to be parsed. Also fixed the error where unbalanced brackets did not throw an error.

### CODE REVIEW

File: config_parser.cc [hunk 1]
```diff
@@ -190,6 +197,7 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
         break;
       }
     } else if (token_type == TOKEN_TYPE_START_BLOCK) {
+      balanced += 1;
```
Comment (nyc) [inline config_parser.cc]: I think it's idiomatic in C++ to say `balanced++;`.

File: config_parser.cc [hunk 2]
```diff
@@ -199,6 +207,7 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
           new_config);
       config_stack.push(new_config);
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
+      balanced += -1;
```
Comment (nyc) [inline config_parser.cc]: Also idiomatic to say `balanced--;` or even `balanced -= 1;` rather than `balanced += -1`.

File: config_parser.cc [hunk 3]
```diff
@@ -211,7 +220,14 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
         // Error.
         break;
       }
+
+      // Check that brackets are balanced
+      if (balanced != 0) {
+        return false;
```
Comment (nyc) [inline config_parser.cc]: Maybe you want to break here instead, like how other errors were handled?

**Comment by nyc** [comment]
Looks good, I left some comments in your commit 83a91fe about some weird uses of `+=`. I also think you could've written a few more tests for simple edge cases like testing that different types of whitespace are parsed correctly (e.g. `\t\r\n`).

**Comment by violet-moon** [comment]
Fixed the code based on the comments and added a unit test for whitespace.
